src/common/stack_mmap.c

```diff
@@ -0,0 +1,422 @@
+/**
+ * (C) Copyright 2016-2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * \file
+ *
+ * This file implements an alternate/external way for ULTs stacks allocation.
+ * It is based on mmap() of MAP_STACK|MAP_GROWSDOWN regions, in order to
+ * allow overrun detection along with automatic growth capability.
+ */
+
+#ifdef ULT_MMAP_STACK
+#include <daos/common.h>
+#include <daos/stack_mmap.h>
+#include <errno.h>
+#include <string.h>
+
+/* ABT_key for mmap()'ed ULT stacks */
+ABT_key stack_key;
+
+/* engine's (ie including all XStreams/stack-pools) max number of mmap()'ed
+ * ULTs stacks, to be based on vm.max_map_count minus an estimate of the
+ * non-stacks mmap()'ed regions required (where malloc() itself will use
+ * mmap() when allocating chunks of size > M_MMAP_THRESHOLD, and there is
+ * a M_MMAP_MAX maximum for such number of chunks, both can be updated
+ * dynamically using mallopt() !!...) for engine operations (including
+ * pre-reqs ...).
+ */
+int max_nb_mmap_stacks;
+
+/* engine's (ie including all XStreams/stack-pools) current number of mmap()'ed
+ * ULTs stacks, to be [in,de]cremented atomically and compared to
+ * max_nb_mmap_stacks
+ */
+ATOMIC int nb_mmap_stacks;
+
+/* engine's (ie including all XStreams/stack-pools) current number of free/queued
+ * mmap()'ed ULTs stacks, to be [in,de]cremented atomically and compared to
+ * max_nb_mmap_stacks
+ */
+ATOMIC int nb_free_stacks;
+
+/* mmap()'ed or Argobot's legacy/internal allocation method for ULT stacks ? */
+bool daos_ult_mmap_stack = true;
+
+/* one per supported ABT_thread_create[_...] API type */
+enum AbtThreadCreateType {
+	MAIN,
+	ON_XSTREAM
+};
+
+static int call_abt_method(void *arg, enum AbtThreadCreateType flag,
+			   void (*thread_func)(void *), void *thread_arg,
+			   ABT_thread_attr attr, ABT_thread *newthread)
```

Comment:
Should be:
static in
call_abt_method(..)
{
...

Same for all functions in this file.